### DISCUSSION TITLE
Allow only specific people or team to approve a PR

### DISCUSSION THREAD

**Discussion by mantzas** [community/community#23765] [New to GitHub]
<p>Hi,</p>
<p>we have private repos and we are allowing other people to have write access to the repo.</p>
<p>We want to be able to restrict though who is allowed to approve a PR based on member or even better a team.</p>
<p>We want also to have at least 3 people of the owning team to give approvals before the merge.</p>
<p>So, in essence, we want to differentiate between contributing teams and approving teams.</p>
<p>I know that in Gitlab this is possible, I assume in Github also but I cannot find this setting anywhere.</p>
<p>Code owners are restricted to have at least one of the approvers to be an owner but we want more…</p>
<p>Any idea?</p>

**Comment by ghost** [comment, marked as answer]
<p>Hi <a class="mention" href="/mantzas">@mantzas</a>! Thanks for your question.</p>
<p>Through <a href="https://help.github.com/en/github/administering-a-repository/enabling-required-reviews-for-pull-requests">protected branch settings</a>, you can specify a minimum number of reviews from collaborators with write or admin permissions on the repo.</p>
<p>You can combine this with a <a href="https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners">codeowners file</a> to ensure that at least one member of a specific team approves changes for code that they own. However it’s not possible to require that more than one member of the team in the codeowners file approves the change.</p>
<p>I hope this helps!</p>

**Reply by swrobel** [reply]
I just want to verify that this setting actually requires a review from a collaborator with write or admin permissions, because according to [this page](https://docs.github.com/en/organizations/managing-access-to-your-organizations-repositories/repository-roles-for-an-organization#permissions-for-each-role), the "read" permission can submit PR reviews, which seems 🤔 to me...
> Require approvals When enabled, pull requests targeting a matching branch require a number of approvals and no changes requested before they can be merged.

**Reply by faizanakram99** [reply]
How is this an answer? We have the same issue and for now we added the approval team as codeowner for whole project but now none of the other codeowners get any notification because the first rule matches and all other rules are skipped.

**Reply by divad** [reply]
Gitlab allows, in repo settings, to configure who is allowed to approve PRs on the repo. Github should also have this setting - please consider adding it?

**Reply by Alex-ley-scrub** [reply]
Azure DevOps also has a very nice solution to this you can say the PR requires X approvals from Y group of users and Y group can be whatever you want - we defined a group for that purpose. You can also say it requires X from group Y and N from group Z and M approvals overall etc. This is lacking a lot in GitHub and is a pain.

**Reply by OddKMS** [reply]
@Alex-ley-scrub Aye for some reason I expected this functionality to already be present in GitHub since Azure DevOps often lags behind its competetors - especially considering how long GitHub has been around.

**Reply by dieter-aerit** [reply]
Just want to add in that this isn't a good answer.
I want to be able to configure that team(s) A  and B need to approve a PR before it can go out.

Since nowadays a lot of process control happens through VCS, being able to enforce stakeholder approval is pretty important.

**Reply by pdefreitas** [reply]
We had on ADO similar setup to what @Alex-ley-scrub mentioned. We would like to have a similar feature.

**Reply by AwolDes** [reply]
+1 to the replies here. It is limiting in our org because we encourage people in non-technical roles (i.e Support, Product Managers) to write small PRs to update copy or even write small features. Restricting their ability to approve of PRs would be very useful, as it means an approval comes with sign off that the technical quality meets our standards. A similar use case would be team members who mostly QA & have write access for copy updates.

**Reply by bioche** [reply]
Github has added rulesets but weirdly it still doesn't handle this common use case :/

**Comment by akashdip2001** [comment]
GitHub does provide features to control code review and approvals based on teams and individuals. To achieve what you're looking for, you can use branch protection rules combined with code review assignment.

Here's a guide:

1. **Teams:**
   - Ensure that you have teams created on your GitHub organization that represent contributing teams and approving teams.

2. **Branch Protection:**
   - Go to your repository on GitHub.
   - Navigate to "Settings" > "Branches."
   - Add a branch protection rule for the branch you want to protect (e.g., master).
   - Enable "Require pull request reviews before merging."

3. **Reviewers:**
   - Under "Require review from Code Owners," specify the team(s) that should be responsible for code reviews.
   - Add other teams or individuals who should also review the code.

4. **Required Reviewers:**
   - Enable "Dismiss stale pull request approvals when new commits are pushed."
   - Set the "Required approving review" to 3 (or the number you prefer).

By configuring these settings, you can ensure that specific teams or individuals are required to review the code, and you can set the number of approvals needed before merging. Keep in mind that GitHub doesn't directly have a concept of "approvers" distinct from code owners, but you can achieve the desired workflow through the settings mentioned above.

## Profile
![Profile Image](https://thumbnails.odycdn.com/optimize/s:200:0/quality:95/plain/https://thumbs.odycdn.com/bf58cf4d7ec52c62f02c52e16f150a5b.webp)
**AKASHDIP MAHAPATRA**
"everything is physics bro"  
E=mc²

|[Subscribe on YouTube](https://www.youtube.com/channel/UCxvmp634YDc41xCWOdvWqoQ?sub_confirmation=1) |
|----  |

**Reply by dieter-aerit** [reply]
@akashdip2001 : But as I understand it works now, if my code owners mentions 2 groups, let's call them "contributors" and "approvers", and I require X number of approvers on the PR, then GitHub would be fine with X contributors **OR** X approvers giving their approval on the PR.

I want to enforce that both contributors **AND** approvers give their consent on the PR, and that is not possible as far as I know?

So if I require 2 reviewers, I want it to be a contributor and an approver. NOT 2 contributors or 2 approvers.

**Reply by akashdip2001** [reply]
To achieve the requirement where you want both "contributors" and "approvers" to give their consent on a pull request (PR), you can't rely on the basic 'required reviewers' functionality provided by GitHub as it does not differentiate between groups of reviewers. However, there are workarounds to impose such constraints:

1. **Pull Request Templates**: Use a PR template to instruct contributors that they must have approval from at least one member of each group. This is an honor system and relies on individuals adhering to the guidelines.

2. **CODEOWNERS File**: More granular control can be achieved by using the `CODEOWNERS` file. You can segment the codebase into sections and assign a required approver (or approvers) for each section. You could theoretically set this up to enforce separate approvals from contributors and approvers.

   However, enforcing a strict separation where exactly one of each must approve is still not achievable with this alone.

3. **Status Checks**: You can implement a custom status check using GitHub Actions or external CI/CD services. This custom script would check the list of reviewers and only pass if it includes the required mix of contributors and approvers.

4. **Protected Branches with Required Status Checks**: Use branch protection rules to require status checks to pass before merging. Combining this with the custom status check script from point 3 can enforce your requirement.

5. **GitHub Apps or Bots**: Create or use a GitHub App or a bot that can block a PR merge if the approval criteria are not met. The app would programmatically evaluate the current approvals and compare them against your criteria (one contributor and one approver).

6. **Review Required Labels**: Use labels and GitHub Actions to enforce that a contributor's approval adds a specific label (say, `contributor-approved`) and an approver's approval adds another (`approver-approved`). A PR can only be merged if it has both labels.

Remember, automation solutions like GitHub Actions, bots, or apps will require additional maintenance and oversight but can provide the level of control you're looking for.

While the custom automation setup might require initial investment in time and possibly infrastructure (if you don't use existing GitHub Actions or Apps), it is the most reliable way to enforce complex review policies.

**Reply by mithunpandey** [reply]
The Use case is very clear,
How to enforce that Unless one reviewer each from Multiple Teams mentioned in the CODEOWNERS file for single Pull Request approves it the PR cant be merged.

IT seems like this can not be done by using the Branch Protection settings and CODEOWNERS file

**Reply by nathannis** [reply]
Internally we have a similar policy (ours is one developer familiar with the source code AND one generic developer) I initially set up our code owners file to have both GitHub teams (@silo-team and @developers) but then eventually realized that if two @developers but no @silo-team members approved the button turned green on GitHub and a developer from @silo-team was no longer required to approve prior to merging. We eventually set up our own app to assign the additional @developer and only put @silo-team in the github code owners file.

**Comment by Shreyas-Jangam** [comment]
Enable Branch Protection Rules
Go to your repository on GitHub.
Navigate to Settings > Branches > Branch protection rules.
Click on Add rule and specify the branch you want to protect (e.g., main).
Under Branch protection rules, enable Require pull request reviews before merging.
Enable Require review from Code Owners to ensure only code owners can approve.
2. Define Code Owners
In the root directory of your repository, create a .github/CODEOWNERS file (if it doesn’t exist already).
Specify the paths and the users or teams responsible. For example:
plaintext
Copy code
# This makes only the specified users or team members code owners for the repository
* @username1 @username2
# Or specify a team
* @org/team-name
Save and commit the .github/CODEOWNERS file.
With this setup, only the specified users or teams in the CODEOWNERS file can approve the PRs. This is particularly useful in repositories where approvals are required only from particular team members or domain experts.

**Comment by LeoLau94** [comment]
Hi，Can I achieve it without a CODEOWNERS file?

**Reply by sandya-kandregula** [reply]
How can we make it happen that approval from each team 
Example approval should be done by team A and team B as well 
So min 2 approvals from each team.
